metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/sql/DirectSqlGetPartition.java

```diff
@@ -654,42 +790,154 @@ private String prepareFilterSql(@Nullable final String filterExpression) {
         if (sort != null && sort.hasSort()) {
             queryBuilder.append(" order by ").append(sort.getSortBy()).append(" ").append(sort.getOrder().name());
         }
+        //TODO: check if this is needed. Why don't we allow pageable without filter
         if (pageable != null && pageable.isPageable() && Strings.isNullOrEmpty(filterExpression)) {
             if (sort == null || !sort.hasSort()) {
                 queryBuilder.append(" order by p.part_id");
             }
             queryBuilder.append(" limit ").append(pageable.getOffset()).append(',').append(pageable.getLimit());
         }
 
-        List<T> partitions = Lists.newArrayList();
+        List<T> partitions;
         final List<Object> params = Lists.newArrayList(databaseName, tableName);
         if (filterSql != null && filterParams != null) {
             params.addAll(filterParams);
         }
         final Object[] oParams = new Object[params.size()];
         partitions = (List) jdbcTemplate.query(
             queryBuilder.toString(), params.toArray(oParams), resultSetExtractor);
-        //
+
+        //TODO: check why processing pageable twice
         if (pageable != null && pageable.isPageable() && !Strings.isNullOrEmpty(filterExpression)) {
-            int limit = pageable.getOffset() + pageable.getLimit();
-            if (partitions.size() < limit) {
-                limit = partitions.size();
+            partitions = processPageable(partitions, pageable);
+        }
+        return partitions;
+    }
+
+    /**
+     * Check if a WAP pattern table, i.e. teh database is audit and the table name matches WAP table pattern
+     *
+     * @param databaseName database
+     * @param tableName    table name
+     * @return true or false
+     */
+    private boolean isWAPAuditTable(final String databaseName, final String tableName) {
```

Comment:
Can we have this method return an Optional<QualifiedName>?

Reply:
Agreed.